### PR TITLE
Truffle deploy

### PR DESCRIPTION
**Pull request by Siegrift**
You need to have `truffle` installed (preferably globally), then you can use `truffle migrate --network ropsten` to deploy to testnet or `truffle migrate` to deploy to local ganache.
 
UPDATE:
There was more handling required to make this fully work (and also a few hacks). Notably:
- I previously forgot about our way of linking contracts by modifying bytecode - this was implemented already in [memefactory utils](https://github.com/district0x/memefactory/blob/59c56693fc8a5750f229d2e29a9567352e986d0b/migrations/utils.js#L54)
- Once the bytecode is replaced it is actually persisted to file system. Combined with how truffle only compiles contracts if there are changes in the source files you basically get 5 hours of slow debugging why testnet deploy doesn't work and using ganache works. I avoided that by making sure the contracts are always recompiled.
- I also fixed the tests, since they were broken after wallet connect integration.
- Polished `project.clj` - The transitive dependency issue I experienced was probably just some inconsistency in the project at that time.

### CODE REVIEW

File: src/name_bazaar/server/dev.cljs [hunk 1]
```diff
@@ -72,16 +72,16 @@
                                       :console? true}
                             :endpoints {:port 6200
                                         :middlewares [logging-middlewares]}
-                            :web3 {:port 8549}
+                            :web3 {:url "https://ropsten.infura.io/"}
```
Comment (Siegrift) [inline src/name_bazaar/server/dev.cljs]: This should actually be `https://ropsten.infura.io/v3/project-id`

File: src/name_bazaar/ui/config.cljs [hunk 2]
```diff
@@ -7,7 +7,7 @@
    :logging {:level :debug
              :console? true}
    :pushroute-hosts "localhost"
-   :node-url "http://localhost:8549"
+   :node-url "https://ropsten.infura.io/"
```
Comment (Siegrift) [inline src/name_bazaar/ui/config.cljs]: This should actually be `https://ropsten.infura.io/v3/project-id`

File: README.md [hunk 3]
```diff
@@ -50,6 +50,10 @@ folder and try to start the server again)_
 
 #### Redeploy smart-contracts and generate mock data
 
+TODO: update this section to use truffle and remove redeploy function from clojure
+
+TODO: to run deploy on testnet: `truffle migrate --network ropsten`
```
Comment (ladislavdubravsky) [inline README.md]: There's Deploying Name Bazaar smart contracts section below, you can add a Truffle deploy subsection there as the first option, but keep the clojure one as fallback.

Comment (Siegrift) [reply README.md]: You are commenting on outdated files (don't know how :D)

File: resources/public/contracts/src/NameBazaarDevContracts.sol [hunk 4]
```diff
@@ -0,0 +1,23 @@
+pragma solidity ^0.5.17;
+pragma experimental ABIEncoderV2;
+
+// TODO: merge with NameBazaarRegistrar into single "development only" file
+import {PublicResolver} from "@ensdomains/resolver/contracts/PublicResolver.sol";
+import {ReverseRegistrar, NameResolver} from "@ensdomains/ens/contracts/ReverseRegistrar.sol";
+import {ENS} from "@ensdomains/ens/contracts/ENS.sol";
+
+contract NamebazaarDevPublicResolver is PublicResolver {
```
Comment (ladislavdubravsky) [inline resources/public/contracts/src/NameBazaarDevContracts.sol:12]: :+1:

File: README.md (hunk 3 above)
Comment (ladislavdubravsky) [reply README.md]: I forgot I only clicked the first commit :D

File: src/name_bazaar/server/deployer.cljs [hunk 5]
```diff
@@ -30,11 +30,22 @@
...
-(defn deploy-registrar! [default-opts]
+(defn deploy-resolvers! [default-opts]
+  (deploy-smart-contract! :reverse-name-resolver (merge default-opts {:gas 3200000}))
+  (deploy-smart-contract! :public-resolver (merge default-opts
+                                                    {:gas 3200000
+                                                     :arguments [(contract-address :ens)]})))
+
+
+(defn deploy-registrars! [default-opts]
   (deploy-smart-contract! :name-bazaar-registrar (merge default-opts
```
Comment (ladislavdubravsky) [inline src/name_bazaar/server/deployer.cljs]: `:name-bazaar-registrar` should really be called `eth-registrar` everywhere, even in derived function names. It was called `:registrar` when we started, but we managed to rename it wrong :)

File: src/name_bazaar/ui/events/ens_events.cljs [hunk 6]
```diff
@@ -66,7 +66,7 @@
                                                         :fields [:offering/node :offering/address]
                                                         :node-owner? true}
                                                :on-success [:ens.records.active-offerings/loaded nodes]
-                                               :on-failure [::logging/error "Failed to load END record active offerings"
```
Comment (ladislavdubravsky) [inline src/name_bazaar/ui/events/ens_events.cljs:69]: :smile:

File: project.clj [hunk 7]
```diff
@@ -96,19 +95,7 @@
                        ["@ensdomains/ethregistrar" "2.0.4"]
                        ["@ensdomains/resolver" "0.2.4"]
                        ["@sentry/node" "4.2.1"]
-                       ;; https://github.com/district0x/district-server-db/blob/3839edd/project.clj#L12
-                       ;; After deployment changes, (transitive) dependencies are not installed automatically
-                       ;; TODO: remove, they should be installed automatically
-                       [better-sqlite3 "5.4.0"]
                        [chalk "2.3.0"]
-                       ;; https://github.com/district0x/district-server-smart-contracts/blob/682e649/project.clj#L17
```
Comment (ladislavdubravsky) [inline project.clj]: These don't get installed for me on clean install, so I seem to need the dependencies

File: README.md [hunk 8]
```diff
@@ -24,6 +24,8 @@ Open another terminal, start autocompiling smart-contracts
 
 ```bash
 lein auto compile-solidity
+# or alternatively use (note: truffle is currently experimental)
+truffle compile
```
Comment (ladislavdubravsky) [inline README.md]: Starting dev server in this workflow, on `node dev-server/name-bazaar.js` I get 

```could not start [#'district.server.smart-contracts/smart-contracts] due to Error: ENOENT: no such file or directory, open '/home/ladislav/prj/NameBazaar/name-bazaar/resources/public/contracts/build/AuctionOfferingFactory.abi'```

and indeed the old directory doesn't contain build now.

Comment (Siegrift) [reply README.md]: Oh, my bad. Yeah, currently truffle compile is not a replacement for `lein compile`.

It compiles the contracts, but only the "truffle way" and the clojure(script) layer needs the ABI's

File: project.clj (hunk 7 above)
Comment (Siegrift) [reply project.clj]: :/ Ok, maybe I had package.lock or something preserved, but I thought I was doing a clean install as well...

File: src/name_bazaar/server/deployer.cljs (hunk 5 above)
Comment (Siegrift) [reply src/name_bazaar/server/deployer.cljs]: True, lets rename it everywhere in code.

**Review by ladislavdubravsky — APPROVED** [review]
LGTM!

**Pull request merged**